pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +266,178 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+}
+
+// validateCSIDriverInfoStatusName tests if Name in CSIDriverInfoStatus is a valid name.
+func validateCSIDriverInfoStatusName(name string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, name))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecNodeID tests if Name in CSIDriverInfoSpec is a valid node id.
+func validateCSIDriverInfoSpecNodeID(nodeID string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(nodeID) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, nodeID))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecTopologyKeys tests if TopologyKeys in CSIDriverInfoSpec are
+// valid topology keys.
+func validateCSIDriverInfoSpecTopologyKeys(keys []string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(keys) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	}
+	for _, key := range keys {
+		if len(key) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, key))
+		}
+	}
+	return allErrs
+}
+
+var supportedVolumePluginMechanisms = sets.NewString(string("in-tree"), string("csi"))
+
+// validateCSIDriverInfoStatusVolumePluginMechanism tests if VolumePluginMechanism in
+// CSIDriverInfoStatus is valid
+func validateCSIDriverInfoStatusVolumePluginMechanism(volumePluginMechanism string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !supportedVolumePluginMechanisms.Has(volumePluginMechanism) {
+		allErrs = append(allErrs, field.NotSupported(fldPath, volumePluginMechanism, supportedVolumePluginMechanisms.List()))
+	}
+
+	return allErrs
+}
+
+// ValidateCSIDriver validates a CSIDriver.
+func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&csiDriver.ObjectMeta, false, apivalidation.ValidateClassName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateCSIDriverSpec(&csiDriver.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateCSIDriverUpdate validates a CSIDriver.
+func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
+	allErrs := ValidateCSIDriver(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	}
+	return allErrs
+}
+
+// ValidateCSIDriverSpec tests that the specified CSIDriverSpec
+// has valid data.
+func validateCSIDriverSpec(
+	spec *storage.CSIDriverSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validatePodInfoOnMountVersion(spec.PodInfoOnMountVersion, fldPath.Child("podInfoOnMountVersion"))...)
+	return allErrs
+}
+
+var supportedPodInfoOnMountVersions = sets.NewString(string("v1"))
```

Comment:
nit: global var in top block. add comment for ones that need explaination (I think this one does).

Ditto for any other global vars declared

Reply:
Done